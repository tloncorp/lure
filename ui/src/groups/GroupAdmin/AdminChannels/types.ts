import { GroupChannel } from '@/types/groups';

export interface SectionMap {
  [key: string]: SectionListItem;
}

export interface ChannelListItem {
  key: string;
  channel: GroupChannel;
}

export interface SectionListItem {
  title: string;
  isNew?: boolean;
  channels: ChannelListItem[];
}
