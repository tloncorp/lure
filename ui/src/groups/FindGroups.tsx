import React, { useCallback, useEffect, useState } from 'react';
import { Helmet } from 'react-helmet';
import cn from 'classnames';
import ob from 'urbit-ob';
import {
  useGangs,
  useGroupState,
  usePendingGangsWithoutClaim,
} from '@/state/groups';
import { useIsMobile } from '@/logic/useMedia';
import ShipSelector, { ShipOption } from '@/components/ShipSelector';
import { Gangs, GroupIndex, ViewProps } from '@/types/groups';
import useRequestState from '@/logic/useRequestState';
import { hasKeys, preSig, whomIsFlag } from '@/logic/utils';
import { useNavigate, useParams, useLocation } from 'react-router';
import asyncCallWithTimeout from '@/logic/asyncWithTimeout';
import { useModalNavigate } from '@/logic/routing';
import GroupJoinList from './GroupJoinList';
import GroupJoinListPlaceholder from './GroupJoinListPlaceholder';

export default function FindGroups({ title }: ViewProps) {
  const { ship, name } = useParams<{ ship: string; name: string }>();
  const location = useLocation();
  const navigate = useNavigate();
  const modalNavigate = useModalNavigate();
  const [groupIndex, setGroupIndex] = useState<GroupIndex | null>(null);
  const existingGangs = useGangs();
  const pendingGangs = usePendingGangsWithoutClaim();
  const isMobile = useIsMobile();

  /**
   *  Search results for render:
   *
   *  - Filter out index results for gangs with invites, since they will be
   *    displayed in the Pending Invites section below.
   *  - Only show Public and Private groups, unless already a member of a ship's
   *    Secret group
   * */
  const indexedGangs = groupIndex
    ? Object.entries(groupIndex)
        .filter(([flag, preview]) => {
          if (flag in existingGangs) {
            // Bucket invites separately (in Pending Invites section)
            if (existingGangs[flag].invite) {
              return false;
            }
            // Show group if already a member
            return true;
          }

          // Hide secret gangs
          return !('afar' in preview.cordon);
        })
        .reduce((memo, [flag, preview]) => {
          // Invite URL case: only show the linked group
          if (name) {
            return flag === preSig(`${ship}/${name}`)
              ? {
                  ...memo,
                  [flag]: {
                    preview,
                    invite: null,
                    claim:
                      flag in existingGangs ? existingGangs[flag].claim : null,
                  },
                }
              : memo;
          }

          // Otherwise, show all indexed groups
          return {
            ...memo,
            [flag]: {
              preview,
              invite: null,
              claim: flag in existingGangs ? existingGangs[flag].claim : null,
            },
          };
        }, {} as Gangs)
    : null;

  useEffect(() => {
    if (indexedGangs && hasKeys(indexedGangs)) {
      const indexedFlags = Object.keys(indexedGangs);
      if (indexedFlags.every((f) => f in existingGangs)) {
        // The gangs state has already been merged with the indexed gangs,
        // so no need to update again
        return;
      }
      useGroupState.setState((draft) => ({
        gangs: { ...draft.gangs, ...indexedGangs },
      }));
    }
  }, [existingGangs, indexedGangs]);

  const [shipSelectorShips, setShipSelectorShips] = useState<ShipOption[]>([]);

  const selectedShip =
    shipSelectorShips.length > 0 ? shipSelectorShips[0] : null;
  const presentedShip = selectedShip
    ? selectedShip.label || selectedShip.value
    : '';
  const { isPending, setPending, setReady } = useRequestState();

  const searchGroups = useCallback(async () => {
    if (!ship) {
      return;
    }

    setGroupIndex(null);
    setPending();
    try {
      // @ts-expect-error results will always either be a GroupIndex, or the
      // request will throw an error, which will be caught below
      const results: GroupIndex = await asyncCallWithTimeout(
        useGroupState.getState().index(preSig(ship)),
        10 * 1000
      );
      setGroupIndex(results);
      setReady();
    } catch (error) {
      console.log(
        '[FindGroups:SearchGroups] Request failed due to timeout or network issue'
      );
      setGroupIndex({});
      setReady(); // TODO: show error state? e.g. request timed out... or "Is the host online?"
    }
  }, [setPending, setReady, ship]);

  // if ship in query params, do query
  useEffect(() => {
    if (!ship) {
      return;
    }

    searchGroups();
  }, [ship, searchGroups]);

  // once a ship is selected, redirect to find/[selected query]
  useEffect(() => {
    if (selectedShip) {
      navigate(`/find/${selectedShip.value}`);
      return;
    }

    if (ship && name) {
      navigate(`/find/${ship}/${name}`);
      return;
    }

    // user has cleared selection, redirect back to find root
    navigate('/find');
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [selectedShip]);

  const handleProfileClick = () => {
    modalNavigate(`/profile/${preSig(ship!)}`, {
      state: { backgroundLocation: location },
    });
  };

  const resultsTitle = () => {
    if (isPending) {
      return (
        <>
          <span>Searching for groups hosted by&nbsp;</span>
          <span
            onClick={handleProfileClick}
            className="cursor-pointer text-gray-800"
          >
            {presentedShip === '' ? ship : presentedShip}
          </span>
          <span>&nbsp;...</span>
        </>
      );
    }

    if (indexedGangs) {
      if (hasKeys(indexedGangs)) {
        return (
          <>
            <span>Groups hosted by&nbsp;</span>
            <span
              onClick={handleProfileClick}
              className="cursor-pointer text-gray-800"
            >
              {presentedShip === '' ? ship : presentedShip}
            </span>
            <span>:</span>
          </>
        );
      }

      return <span>This ship doesn&apos;t host any groups</span>;
    }

    return null;
  };

  // Allow selecting a ship name or invite URL (i.e., flag) in ShipSelector
  const isValidNewOption = (val: string) =>
    val ? ob.isValidPatp(preSig(val)) || whomIsFlag(val) : false;

  return (
    <>
      {isMobile && (
        <header className="flex h-14 items-center justify-between px-5 py-4">
          <h1 className="text-base font-bold">Find Groups</h1>
        </header>
      )}
      <div
        className={cn('flex grow overflow-y-auto', !isMobile && 'bg-gray-50')}
      >
        <Helmet>
          <title>{title ? title : document.title}</title>
        </Helmet>
        <div className="w-full p-4">
          <section
            className={cn('mb-8 space-y-8', !isMobile && 'card mb-4 sm:p-8')}
          >
            {!isMobile && <h1 className="text-lg font-bold">Find Groups</h1>}
            <div>
              <label htmlFor="flag" className="mb-1.5 block font-semibold">
                Join Groups via Nickname or Urbit ID
              </label>
              <div className="flex flex-col space-y-2">
                <ShipSelector
                  ships={shipSelectorShips}
                  setShips={setShipSelectorShips}
                  isMulti={false}
                  isClearable={true}
                  isLoading={isPending}
                  hasPrompt={false}
                  placeholder={''}
                  isValidNewOption={isValidNewOption}
                />
              </div>
            </div>
            {selectedShip || (ship && name) ? (
              <section className="space-y-3">
                <p className="font-semibold text-gray-400">{resultsTitle()}</p>
                {isPending ? (
                  <GroupJoinListPlaceholder />
                ) : indexedGangs && hasKeys(indexedGangs) ? (
                  <GroupJoinList gangs={indexedGangs} />
                ) : null}
              </section>
            ) : null}
          </section>
          {hasKeys(pendingGangs) ? (
            <section
              className={cn(
                'mb-4 space-y-4',
                !isMobile && 'card space-y-8 sm:p-8'
              )}
            >
              <h1
                className={cn('font-bold', isMobile ? 'text-base' : 'text-lg')}
              >
                Pending Invites
              </h1>
              <GroupJoinList gangs={pendingGangs} />
            </section>
          ) : null}
        </div>
      </div>
    </>
  );
}
