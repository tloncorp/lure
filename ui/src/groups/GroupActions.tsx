import cn from 'classnames';
import React, { PropsWithChildren, useCallback, useState } from 'react';
import * as DropdownMenu from '@radix-ui/react-dropdown-menu';
import { Link, useLocation } from 'react-router-dom';
import InviteIcon16 from '@/components/icons/InviteIcon16';
import LinkIcon16 from '@/components/icons/LinkIcon16';
import PinIcon16 from '@/components/icons/PinIcon16';
import Person16Icon from '@/components/icons/Person16Icon';
import EllipsisIcon from '@/components/icons/EllipsisIcon';
import BulletIcon from '@/components/icons/BulletIcon';
import { useChatState, usePinnedGroups } from '@/state/chat';
import LeaveIcon from '@/components/icons/LeaveIcon';
import useIsGroupUnread from '@/logic/useIsGroupUnread';
import UnreadIndicator from '@/components/Sidebar/UnreadIndicator';
import { citeToPath, useCopy } from '@/logic/utils';
import { useGroupInviteUrl, useLureEnabled } from '@/state/lure/lure';

const { ship } = window;

export function useGroupActions(flag: string) {
  const { doCopy } = useCopy(citeToPath({ group: flag }));
  const [isOpen, setIsOpen] = useState(false);
  const [copyItemText, setCopyItemText] = useState('Copy Group Link');
  const pinned = usePinnedGroups();
  const isPinned = Object.keys(pinned).includes(flag);
  const [groupInviteUrl] = useGroupInviteUrl(flag);
  const [groupInviteUrlText, setGroupInviteUrlText] =
    useState('Copy Invite URL');
  const doCopyInviteUrlText = useCopy(groupInviteUrl).doCopy;
  const [lureEnabled] = useLureEnabled(flag);

  const onCopy = useCallback(() => {
    doCopy();
    setCopyItemText('Copied!');
    setTimeout(() => {
      setCopyItemText('Copy Group Link');
      setIsOpen(false);
    }, 2000);
  }, [doCopy]);

  const onInviteUrl = useCallback(() => {
    doCopyInviteUrlText();
    setGroupInviteUrlText('Copied!');
    setTimeout(() => {
      setCopyItemText('Copy Invite URL');
      setIsOpen(false);
    }, 2000);
  }, [doCopyInviteUrlText]);

  const onPinClick = useCallback(
    // eslint-disable-next-line prefer-arrow-callback
    function <T>(e: React.MouseEvent<T>) {
      e.stopPropagation();
      useChatState.getState().togglePin(flag, !isPinned);
    },
    [flag, isPinned]
  );

  return {
    isOpen,
    isPinned,
    setIsOpen,
    copyItemText,
    onCopy,
    onPinClick,
    groupInviteUrl,
    groupInviteUrlText,
    onInviteUrl,
    lureEnabled,
  };
}

type GroupActionsProps = PropsWithChildren<{
  flag: string;
  className?: string;
}>;

const GroupActions = React.memo(
  ({ flag, className, children }: GroupActionsProps) => {
    const { isGroupUnread } = useIsGroupUnread();
    const location = useLocation();
    const hasActivity = isGroupUnread(flag);

    const {
      isOpen,
      setIsOpen,
      isPinned,
      copyItemText,
      onCopy,
      onPinClick,
      groupInviteUrlText,
      onInviteUrl,
      groupInviteUrl,
      lureEnabled,
    } = useGroupActions(flag);

    const onInviteUrlSelect = useCallback(
      (event: Event) => {
        event.preventDefault();
        onInviteUrl();
      },
      [onInviteUrl]
    );

    const onCopySelect = useCallback(
      (event: Event) => {
        event.preventDefault();
        onCopy();
      },
      [onCopy]
    );

    return (
      <div className={className}>
        <DropdownMenu.Root open={isOpen} onOpenChange={setIsOpen}>
          <DropdownMenu.Trigger asChild className="appearance-none">
            {children || (
              <div className="relative h-6 w-6">
                {!isOpen && hasActivity ? (
                  <UnreadIndicator
                    className="absolute h-6 w-6 text-blue transition-opacity group-focus-within:opacity-0 group-hover:opacity-0"
                    aria-label="Has Activity"
                  />
                ) : null}
                <button
                  className={cn(
                    'default-focus absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 rounded-lg p-0.5 transition-opacity focus-within:opacity-100 hover:opacity-100 group-focus-within:opacity-100 group-hover:opacity-100',
                    hasActivity && 'text-blue',
                    isOpen ? 'opacity:100' : 'opacity-0'
                  )}
                  aria-label="Open Group Options"
                >
                  <EllipsisIcon className="h-6 w-6" />
                </button>
              </div>
            )}
          </DropdownMenu.Trigger>
          <DropdownMenu.Content className="dropdown min-w-52 text-gray-800">
            <DropdownMenu.Item
              asChild
              className="dropdown-item text-blue hover:bg-blue-soft hover:dark:bg-blue-900"
            >
              <Link
                to={`/groups/${flag}/invite`}
                state={{ backgroundLocation: location }}
                className="flex items-center space-x-2"
              >
                <InviteIcon16 className="h-6 w-6 opacity-60" />
                <span className="pr-2">Invite People</span>
              </Link>
            </DropdownMenu.Item>
            <DropdownMenu.Item
              className={
                'dropdown-item flex items-center space-x-2 text-blue hover:bg-blue-soft hover:dark:bg-blue-900'
              }
              onSelect={onCopySelect}
            >
              <LinkIcon16 className="h-6 w-6 opacity-60" />
              <span className="pr-2">{copyItemText}</span>
            </DropdownMenu.Item>
            <DropdownMenu.Item
              className={`dropdown-item flex items-center space-x-2 text-blue hover:bg-blue-soft hover:dark:bg-blue-900${
                groupInviteUrl === '' || !lureEnabled ? ' hidden' : ''
              }`}
              onSelect={onInviteUrlSelect}
            >
              <LinkIcon16 className="h-6 w-6 opacity-60" />
              <span className="pr-2">{groupInviteUrlText}</span>
            </DropdownMenu.Item>
            <DropdownMenu.Item
              className="dropdown-item flex items-center space-x-2"
              onClick={onPinClick}
            >
              <PinIcon16 className="h-6 w-6 text-gray-600" />
              <span className="pr-2">{isPinned ? 'Unpin' : 'Pin'}</span>
            </DropdownMenu.Item>
            <DropdownMenu.Item asChild className="dropdown-item">
              <Link
                to={`/groups/${flag}/info`}
                // state={{ backgroundLocation: location }}
                className="flex items-center space-x-2"
              >
                <Person16Icon className="m-1 h-4 w-4 text-gray-600" />
                <span className="pr-2">Members &amp; Group Info</span>
              </Link>
            </DropdownMenu.Item>
            {flag.includes(ship) ? null : (
              <DropdownMenu.Item asChild className="dropdown-item">
                <Link
                  to={`/groups/${flag}/leave`}
                  state={{ backgroundLocation: location }}
                  className="flex items-center space-x-2 text-red hover:bg-red-soft hover:dark:bg-red-900"
                >
                  <LeaveIcon className="h-6 w-6" />
                  <span className="pr-2">Leave Group</span>
                </Link>
              </DropdownMenu.Item>
            )}
          </DropdownMenu.Content>
        </DropdownMenu.Root>
      </div>
    );
  }
);

export default GroupActions;
