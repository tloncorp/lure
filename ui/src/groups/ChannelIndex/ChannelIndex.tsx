/* eslint-disable no-console */
import React, { useCallback, useState } from 'react';
import cn from 'classnames';
import { useNavigate } from 'react-router';
import { Link } from 'react-router-dom';
import { Helmet } from 'react-helmet';
import { useRouteGroup, useGroup, useAmAdmin, useVessel } from '@/state/groups';
import { GroupChannel, Zone, ViewProps } from '@/types/groups';
import {
  canReadChannel,
  channelHref,
  getFlagParts,
  isChannelImported,
  isChannelJoined,
  nestToFlag,
} from '@/logic/utils';
import * as DropdownMenu from '@radix-ui/react-dropdown-menu';
import LeaveIcon from '@/components/icons/LeaveIcon';
import BulletIcon from '@/components/icons/BulletIcon';
import { useChatState } from '@/state/chat';
import LoadingSpinner from '@/components/LoadingSpinner/LoadingSpinner';
import CaretDown16Icon from '@/components/icons/CaretDown16Icon';
import PencilIcon from '@/components/icons/PencilIcon';
import useRequestState from '@/logic/useRequestState';
import ChannelIcon from '@/channels/ChannelIcon';
import useChannelSections from '@/logic/useChannelSections';
import { useHeapState } from '@/state/heap/heap';
import { useDiaryState } from '@/state/diary';
import useIsChannelHost from '@/logic/useIsChannelHost';
import useAllBriefs from '@/logic/useAllBriefs';
import { useIsMobile } from '@/logic/useMedia';
import CaretLeft16Icon from '@/components/icons/CaretLeft16Icon';
import MigrationTooltip from '@/components/MigrationTooltip';
import { useStartedMigration } from '@/logic/useMigrationInfo';
import useFilteredSections from '@/logic/useFilteredSections';
import EditChannelModal from '../GroupAdmin/AdminChannels/EditChannelModal';

const UNZONED = 'default';

function GroupChannelRow({
  channel,
  nest,
  migration,
}: {
  nest: string;
  channel: GroupChannel;
  migration: ReturnType<typeof useStartedMigration>;
}) {
  const [_app, flag] = nestToFlag(nest);
  const { ship } = getFlagParts(flag);
  const groupFlag = useRouteGroup();
  const isMobile = useIsMobile();
  const group = useGroup(groupFlag);
  const vessel = useVessel(groupFlag, window.our);
  const briefs = useAllBriefs();
  const { hasStarted, pendingImports } = migration;
  const imported = isChannelImported(nest, pendingImports) && hasStarted(ship);
  const joined = isChannelJoined(nest, briefs);
  const isChannelHost = useIsChannelHost(flag);
  const isAdmin = useAmAdmin(groupFlag);
  const navigate = useNavigate();
  const [editIsOpen, setEditIsOpen] = useState(false);
  const [timer, setTimer] = useState<ReturnType<typeof setTimeout> | null>(
    null
  );
  const { isFailed, isPending, isReady, setFailed, setPending, setReady } =
    useRequestState();
  const join = useCallback(
    (chFlag: string) => {
      const joiner =
        _app === 'chat'
          ? useChatState.getState().joinChat
          : _app === 'heap'
          ? useHeapState.getState().joinHeap
          : useDiaryState.getState().joinDiary;

      joiner(chFlag);
    },
    [_app]
  );
  const leave = useCallback(
    (chFlag: string) => {
      const leaver =
        _app === 'chat'
          ? useChatState.getState().leaveChat
          : _app === 'heap'
          ? useHeapState.getState().leaveHeap
          : useDiaryState.getState().leaveDiary;

      leaver(chFlag);
    },
    [_app]
  );

  const joinChannel = useCallback(async () => {
    try {
      if (timer) {
        clearTimeout(timer);
        setTimer(null);
      }
      setPending();
      await join(flag);
      setReady();
    } catch (error) {
      if (error) {
        console.error(`[ChannelIndex:JoinError] ${error}`);
      }
      setFailed();
      setTimer(
        setTimeout(() => {
          setReady();
          setTimer(null);
        }, 10 * 1000)
      );
    }
  }, [flag, join, setFailed, setPending, setReady, timer]);

  const leaveChannel = useCallback(async () => {
    try {
      leave(flag);
    } catch (error) {
      if (error) {
        console.error(`[ChannelIndex:LeaveError] ${error}`);
      }
    }
  }, [flag, leave]);

  const muteChannel = useCallback(() => {
    // TODO: add channel mute action here
    console.log('mute ...');
  }, []);

  const open = useCallback(() => {
    if (!joined) {
      return;
    }
    navigate(channelHref(groupFlag, nest));
  }, [groupFlag, joined, navigate, nest]);

  const canRead = canReadChannel(channel, vessel, group?.bloc);

  if (!group || !canRead) {
    return null;
  }

  return (
    <li className="my-2 flex flex-row items-center justify-between rounded-lg pl-0 pr-2 hover:bg-gray-50">
      <button
        className={cn(
          'flex w-full items-center justify-start rounded-xl text-left hover:bg-gray-50',
          !isMobile ? 'p-2' : ''
        )}
        onClick={open}
        disabled={!joined || !imported}
      >
        <div className="flex flex-row">
          <div className="mr-3 flex h-12 w-12 items-center justify-center rounded bg-gray-50">
            <ChannelIcon nest={nest} className="h-6 w-6 text-gray-400" />
          </div>
          <div className="flex flex-col justify-evenly">
            {joined && imported && nest ? (
              <Link
                className="font-semibold text-gray-800"
                to={channelHref(groupFlag, nest)}
              >
                {channel.meta.title}
              </Link>
            ) : (
              <div className="font-semibold text-gray-800">
                {channel.meta.title}
              </div>
            )}
          </div>
        </div>
      </button>
      <div className="flex-none">
        {!imported ? (
          <MigrationTooltip ship={ship} side="left">
            <button className="secondary-button" disabled>
              Pending Migration
            </button>
          </MigrationTooltip>
        ) : joined ? (
          <>
            <DropdownMenu.Root>
              <DropdownMenu.Trigger asChild className="appearance-none">
                <button
                  className={cn(
                    'button bg-green-soft text-green mix-blend-multiply dark:bg-green-900 dark:mix-blend-screen hover:dark:bg-green-800',
                    isMobile && 'text-sm'
                  )}
                >
                  <span className="mr-1">Joined</span>{' '}
                  <CaretDown16Icon className="h-4 w-4" />
                </button>
              </DropdownMenu.Trigger>
              <DropdownMenu.Content className="dropdown">
                {/* TODO: un-disable this once channel mutes are complete */}
                <DropdownMenu.Item
                  onSelect={muteChannel}
                  className="dropdown-item disabled flex items-center space-x-2 hover:bg-transparent"
                >
                  <BulletIcon className="h-6 w-6 text-gray-400" />
                  <span className="text-gray-400">Mute Channel</span>
                </DropdownMenu.Item>
                {/* TODO: un-disable this once mentions and mutes are complete */}
                <DropdownMenu.Item
                  onSelect={muteChannel}
                  className="dropdown-item disabled flex items-center space-x-2 hover:bg-transparent"
                >
                  <BulletIcon className="h-6 w-6 text-gray-400" />
                  <span className="text-gray-400">Mute Mentions</span>
                </DropdownMenu.Item>
                {isAdmin ? (
                  <DropdownMenu.Item
                    onSelect={() => setEditIsOpen(!editIsOpen)}
                    className="dropdown-item flex items-center space-x-2"
                  >
                    <PencilIcon className="m-1.5 h-3 w-3 fill-gray-500" />
                    <span>Edit Channel</span>
                  </DropdownMenu.Item>
                ) : null}
                {!isChannelHost ? (
                  <DropdownMenu.Item
                    onSelect={leaveChannel}
                    className="dropdown-item flex items-center space-x-2 text-red hover:bg-red-soft hover:dark:bg-red-900"
                  >
                    <LeaveIcon className="h-6 w-6" />
                    <span>Leave Channel</span>
                  </DropdownMenu.Item>
                ) : null}
              </DropdownMenu.Content>
            </DropdownMenu.Root>
            <EditChannelModal
              editIsOpen={editIsOpen}
              setEditIsOpen={setEditIsOpen}
              nest={nest}
              channel={channel}
            />
          </>
        ) : (
          <button
            disabled={isPending}
            onClick={joinChannel}
            className={cn(
              'button mix-blend-multiply disabled:bg-gray-50 dark:mix-blend-screen',
              {
                'bg-gray-50': isReady || isPending,
                'bg-red-soft': isFailed,
                'text-gray-800': isReady,
                'text-gray-400': isPending,
                'text-red': isFailed,
              },
              isMobile ? 'text-sm' : ''
            )}
          >
            {isPending ? (
              <span className="center-items flex">
                <LoadingSpinner />
                <span className="ml-2">Joining...</span>
              </span>
            ) : isFailed ? (
              'Retry'
            ) : (
              'Join'
            )}
          </button>
        )}
      </div>
    </li>
  );
}

function ChannelSection({
  channels,
  zone,
  migration,
}: {
  channels: [string, GroupChannel][];
  zone: Zone | null;
  migration: ReturnType<typeof useStartedMigration>;
}) {
  const flag = useRouteGroup();
  const group = useGroup(flag);
  const sectionTitle =
    zone && group?.zones && zone in group.zones
      ? group.zones[zone].meta.title
      : '';

  return (
    <>
      {zone !== UNZONED ? (
        <div className="py-4 font-semibold text-gray-400">{sectionTitle}</div>
      ) : null}
      <ul>
        {channels.map(([nest, channel]) => (
          <GroupChannelRow
            nest={nest}
            channel={channel}
            key={channel.added}
            migration={migration}
          />
        ))}
      </ul>
    </>
  );
}

export default function ChannelIndex({ title }: ViewProps) {
  const flag = useRouteGroup();
  const { sectionedChannels } = useChannelSections(flag);
  const filteredSections = useFilteredSections(flag);
  const migration = useStartedMigration(flag);
  const navigate = useNavigate();
  const isAdmin = useAmAdmin(flag);
  const group = useGroup(flag);
  const isMobile = useIsMobile();
  const BackButton = isMobile ? Link : 'div';

  return (
    <section className="w-full sm:overflow-y-scroll">
      <Helmet>
        <title>
          {group ? `All Channels in ${group?.meta?.title} ${title}` : title}
        </title>
      </Helmet>
      <div className="flex flex-row items-center justify-between py-1 px-2 sm:p-4">
        <BackButton
          to={`/groups/${flag}`}
          className={cn(
            'cursor-pointer select-none p-2 sm:cursor-text sm:select-text',
            isMobile && 'flex items-center rounded-lg hover:bg-gray-50'
          )}
          aria-label="Back to Group"
        >
          {isMobile ? (
            <CaretLeft16Icon className="mr-1 h-4 w-4 text-gray-400" />
          ) : null}
          <h1 className="text-base font-semibold sm:text-lg">All Channels</h1>
        </BackButton>
        {isAdmin ? (
          <button
            onClick={() => navigate(`/groups/${flag}/info/channels`)}
            className="rounded-md bg-gray-800 py-1 px-2 text-[12px] font-semibold leading-4 text-white"
          >
            Channel Settings
          </button>
        ) : null}
      </div>
      <div className="p-4">
        {filteredSections.map((section) =>
          sectionedChannels[section] ? (
            <div
              key={section}
              className={cn(
                'mb-2 w-full rounded-xl bg-white py-1',
                !isMobile ? 'pl-4 pr-2' : ''
              )}
            >
              <ChannelSection
                channels={
                  section in sectionedChannels ? sectionedChannels[section] : []
                }
                zone={section}
                migration={migration}
              />
            </div>
          ) : null
        )}
      </div>
    </section>
  );
}
