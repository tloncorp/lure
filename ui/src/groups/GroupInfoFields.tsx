import React, { useEffect, useState } from 'react';
import { useFormContext } from 'react-hook-form';
import GroupInfoPreview from '@/groups/NewGroup/GroupInfoPreview';
import ImageOrColorField, {
  ImageOrColorFieldState,
} from '@/components/ImageOrColorField';
import { useCalm } from '@/state/settings';
import { isValidUrl } from '@/logic/utils';
import { GroupMeta } from '@/types/groups';

export default function GroupInfoFields() {
  const { register, watch } = useFormContext<GroupMeta>();
  const [iconType, setIconType] = useState<ImageOrColorFieldState>('color');
  const [coverType, setCoverType] = useState<ImageOrColorFieldState>('color');
  const watchImage = watch('image');
  const watchTitle = watch('title');
  const watchCover = watch('cover');
  const showEmpty = iconType === 'image' && !isValidUrl(watchImage);
  const showCoverEmpty = coverType === 'image' && !isValidUrl(watchCover);
  const calm = useCalm();

  return (
    <>
      <div className="flex flex-col">
        <label htmlFor="title" className="pb-2 font-bold">
          Group Name*
        </label>
        <input
          // TODO: set sane maxLength
          {...register('title', { required: true, maxLength: 180 })}
          className="input"
          type="text"
          placeholder="e.g. Urbit Fan Club"
        />
      </div>
      <div className="flex flex-col space-x-0 sm:flex-row sm:space-x-2">
        <div className="pb-2 font-bold sm:hidden">Preview</div>
        <div className="flex flex-row justify-center pb-2">
          <GroupInfoPreview
            iconType={iconType}
            showEmpty={showEmpty}
            watchImage={watchImage}
            watchTitle={watchTitle}
            watchCover={watchCover}
            coverType={coverType}
            showCoverEmpty={showCoverEmpty}
          />
        </div>
        <div className="flex grow flex-col">
          <span className="pb-2 font-bold">Group Icon*</span>
          <ImageOrColorField
            fieldName="image"
            state={iconType}
            setState={setIconType}
            defaultColor="#999999"
          />
          <span className="mt-4 pb-2 font-bold">Group Cover*</span>
          <ImageOrColorField
            fieldName="cover"
            state={coverType}
            setState={setCoverType}
            defaultColor="#D9D9D9"
          />
        </div>
      </div>
      <div className="flex flex-col">
        <label htmlFor="description" className="pb-2 font-bold">
          Group Description (optional)
        </label>
        <textarea
          // TODO: set sane maxLength
          {...register('description', { maxLength: 300 })}
          className="input"
          spellCheck={`${!calm.disableSpellcheck}`}
        />
      </div>
    </>
  );
}
