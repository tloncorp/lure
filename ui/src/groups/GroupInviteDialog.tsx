import ob from 'urbit-ob';
import React, { useCallback, useState } from 'react';
import Dialog, { DialogClose, DialogContent } from '@/components/Dialog';
import ShipSelector, { ShipOption } from '@/components/ShipSelector';
import { useDismissNavigate } from '@/logic/routing';
import { useGroup, useGroupState, useRouteGroup } from '@/state/groups/groups';
import { getPrivacyFromGroup, preSig } from '@/logic/utils';

export default function GroupInviteDialog() {
  const dismiss = useDismissNavigate();
  const flag = useRouteGroup();
  const group = useGroup(flag);
  const privacy = group ? getPrivacyFromGroup(group) : 'public';
  const [ships, setShips] = useState<ShipOption[]>([]);
  const validShips = ships
    ? ships.every((ship) => ob.isValidPatp(preSig(ship.value)))
    : false;

  const onOpenChange = (isOpen: boolean) => {
    if (!isOpen) {
      dismiss();
    }
  };

  const onInvite = useCallback(() => {
    const shipList = ships.map((s) => preSig(s.value));
    if (privacy === 'public') {
      useGroupState.getState().addMembers(flag, shipList);
    } else {
      useGroupState.getState().invite(flag, shipList);
    }
  }, [flag, privacy, ships]);

  const onEnter = useCallback(() => {
    onInvite();
    dismiss();
  }, [onInvite, dismiss]);

  return (
    <Dialog defaultOpen onOpenChange={onOpenChange}>
      <DialogContent containerClass="w-full max-w-xl" showClose>
        <div className="flex flex-col">
          <h2 className="mb-4 text-lg font-bold">Invite To Group</h2>
          <div className="w-full py-3">
            <ShipSelector ships={ships} setShips={setShips} onEnter={onEnter} />
          </div>
          <div className="flex items-center justify-end space-x-2">
            <DialogClose className="secondary-button">Cancel</DialogClose>

            <DialogClose
              onClick={onInvite}
              className="button bg-blue text-white dark:text-black"
              disabled={!validShips}
            >
              Invite
            </DialogClose>
          </div>
        </div>
      </DialogContent>
    </Dialog>
  );
}
