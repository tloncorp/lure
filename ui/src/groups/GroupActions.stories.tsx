import React from 'react';
import GroupActions from './GroupActions';

export default {
  component: 'GroupActions',
  title: 'GroupActions',
};

export function Text() {
  return <GroupActions flag={'~zod/test'} />;
}
