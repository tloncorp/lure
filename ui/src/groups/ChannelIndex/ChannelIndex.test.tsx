/* eslint-disable import/no-extraneous-dependencies */
import { createChannel, createMockGroup } from '@/mocks/groups';
import { Group } from '@/types/groups';
import React from 'react';
import { describe, expect, it, vi } from 'vitest';
import { render } from '../../../test/utils';
import ChannelIndex from './ChannelIndex';

const fakeGroup: Group = createMockGroup('Fake Group');
fakeGroup.channels[`chat/~zod/tlon`] = createChannel('Fake Channel');
const fakeVessel = fakeGroup.fleet['~hastuc-dibtux'];

vi.mock('@/state/groups', () => ({
  useGroup: () => fakeGroup,
  useRouteGroup: () => null,
  useAmAdmin: () => true,
  useVessel: () => fakeVessel,
  useGroupState: () => ({}),
}));

describe('ChannelIndex', () => {
  it('renders as expected', () => {
    const { asFragment } = render(<ChannelIndex />);
    expect(asFragment()).toMatchSnapshot();
  });
});
