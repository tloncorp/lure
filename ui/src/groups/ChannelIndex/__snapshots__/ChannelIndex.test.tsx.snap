// Vitest Snapshot v1

exports[`ChannelIndex > renders as expected 1`] = `
<DocumentFragment>
  <section
    class="w-full sm:overflow-y-scroll"
  >
    <div
      class="flex flex-row items-center justify-between py-1 px-2 sm:p-4"
    >
      <div
        aria-label="Back to Group"
        class="cursor-pointer select-none p-2 sm:cursor-text sm:select-text"
        to="/groups/null"
      >
        <h1
          class="text-base font-semibold sm:text-lg"
        >
          All Channels
        </h1>
      </div>
      <button
        class="rounded-md bg-gray-800 py-1 px-2 text-[12px] font-semibold leading-4 text-white"
      >
        Channel Settings
      </button>
    </div>
    <div
      class="p-4"
    />
  </section>
</DocumentFragment>
`;
